primitives/src/circuit/merkle_tree.rs

```diff
@@ -10,26 +10,32 @@
 
 use crate::{
     circuit::rescue::RescueGadget,
-    merkle_tree::{AccMemberWitness, MerklePath, MerkleTree, NodePos, NodeValue},
+    merkle_tree::{
+        examples::RescueMerkleTree, internal::MerkleNode, MerkleTreeScheme, ToTraversalPath,
+    },
     rescue::RescueParameter,
 };
 use ark_ec::TEModelParameters as Parameters;
 use ark_ff::PrimeField;
-use ark_std::{vec, vec::Vec};
+use ark_std::vec::Vec;
 use jf_relation::{errors::CircuitError, BoolVar, Circuit, PlonkCircuit, Variable};
 
-#[derive(Clone)]
-struct MerkleNodeBooleanEncoding<F: PrimeField> {
-    sibling1: NodeValue<F>,
-    sibling2: NodeValue<F>,
+type NodeVal<F> = <RescueMerkleTree<F> as MerkleTreeScheme>::NodeValue;
```

Comment:
Is it possible to make the circuit instantiation more modular so that it doesn't bind to RescueMerkleTree? E.g., suppose we want to support other hashes (e.g. Poseidon or sth else) in the future, it should be easy to do without changing much of the code here.

Reply:
Sure, but maybe we could make this a separate task?
Previously the circuit was also [rescue specific](https://github.com/EspressoSystems/jellyfish/blob/main/primitives/src/circuit/merkle_tree.rs#L198) and I think it requires a bit more refactoring to make it more modular. Namely, the `MerkleTreeGadget` would need to be polymorphic over some kind of `HashGadget` (of which `RescueGadget` would be one instantiation). Unless you see an easier way?

Reply:
https://github.com/EspressoSystems/jellyfish/issues/142